base/compiler/ssair/EscapeAnalysis/EscapeAnalysis.jl

```diff
@@ -0,0 +1,1604 @@
+end
+function getindex(estate::EscapeState, @nospecialize(x))
+    if isa(x, Argument) || isa(x, SSAValue)
+        return estate.escapes[iridx(x, estate)]
+    else
+        return nothing
+    end
+end
+function setindex!(estate::EscapeState, v::EscapeLattice, @nospecialize(x))
+    if isa(x, Argument) || isa(x, SSAValue)
+        estate.escapes[iridx(x, estate)] = v
+    end
+    return estate
+end
+
+"""
+    iridx(x, estate::EscapeState) -> xidx::Union{Int,Nothing}
+
+Tries to convert analyzable IR element `x::Union{Argument,SSAValue}` to
+its unique identifier number `xidx` that is valid in the analysis context of `estate`.
+Returns `nothing` if `x` isn't maintained by `estate` and thus unanalyzable (e.g. `x::GlobalRef`).
+
+`irval` is the inverse function of `iridx` (not formally), i.e.
+`irval(iridx(x::Union{Argument,SSAValue}, state), state) === x`.
+"""
+function iridx(@nospecialize(x), estate::EscapeState)
+    if isa(x, Argument)
+        xidx = x.n
+        @assert 1 ≤ xidx ≤ estate.nargs "invalid Argument"
+    elseif isa(x, SSAValue)
+        xidx = x.id + estate.nargs
+    else
+        return nothing
+    end
+    return xidx
+end
+
+"""
+    irval(xidx::Int, estate::EscapeState) -> x::Union{Argument,SSAValue}
+
+Converts its unique identifier number `xidx` to the original IR element `x::Union{Argument,SSAValue}`
+that is analyzable in the context of `estate`.
+
+`iridx` is the inverse function of `irval` (not formally), i.e.
+`iridx(irval(xidx, state), state) === xidx`.
+"""
+function irval(xidx::Int, estate::EscapeState)
+    x = xidx > estate.nargs ? SSAValue(xidx-estate.nargs) : Argument(xidx)
+    return x
+end
+
+function getaliases(x::Union{Argument,SSAValue}, estate::EscapeState)
+    xidx = iridx(x, estate)
+    aliases = getaliases(xidx, estate)
+    aliases === nothing && return nothing
+    return Union{Argument,SSAValue}[irval(aidx, estate) for aidx in aliases]
+end
+function getaliases(xidx::Int, estate::EscapeState)
+    aliasset = estate.aliasset
+    root = find_root!(aliasset, xidx)
+    if xidx ≠ root || aliasset.ranks[xidx] > 0
+        # the size of this alias set containing `key` is larger than 1,
+        # collect the entire alias set
+        aliases = Int[]
+        for aidx in 1:length(aliasset.parents)
+            if aliasset.parents[aidx] == root
+                push!(aliases, aidx)
+            end
+        end
+        return aliases
+    else
+        return nothing
+    end
+end
+
+isaliased(x::Union{Argument,SSAValue}, y::Union{Argument,SSAValue}, estate::EscapeState) =
+    isaliased(iridx(x, estate), iridx(y, estate), estate)
+isaliased(xidx::Int, yidx::Int, estate::EscapeState) =
+    in_same_set(estate.aliasset, xidx, yidx)
+
+"""
+    EscapeLatticeCache(x::EscapeLattice) -> x′::EscapeLatticeCache
+
+The data structure for caching `x::EscapeLattice` for interprocedural propagation,
+which is slightly more efficient than the original `x` object.
+"""
+struct EscapeLatticeCache
```

Comment:
This should be named something that indicates it describes the escape properties of an argument. It can also move to Core so it can be stored in CodeInstance. Maybe something like `ArgumentEscapeInfo`?

Reply:
Sounds good, I renamed it to `ArgEscapeInfo`.